tensorboard/data/provider.py

```diff
@@ -0,0 +1,223 @@
+        the data to be queried. Required.
+      downsample_to: Integer number of steps to which to downsample the
+        results (e.g., `1000`). Required.
+      run_tag_filter: Optional `RunTagFilter` value. If provided, a time
+        series will only be included in the result if its run and tag
+        both pass this filter. If `None`, all time series will be
+        included.
+      step_filter: Optional `StepFilter` value. If `None`, the entire
+        range of steps may be included.
+
+    The result will only contain keys for run-tag combinations that
+    actually exist, which may not include all entries in the
+    `run_tag_filter`.
+
+    Returns:
+      A nested map `d` such that `d[run][tag]` is a list of
+      `ScalarDatum` values sorted by step.
+    """
+    pass
+
+  def list_tensors(self):
+    """Not yet specified."""
+    pass
+
+  def read_tensors(self):
+    """Not yet specified."""
+    pass
+
+  def list_blob_sequences(self):
+    """Not yet specified."""
+    pass
+
+  def read_blob_sequences(self):
+    """Not yet specified."""
+    pass
+
+
+_ScalarMetadata = collections.namedtuple(
+    "ScalarMetadata",
+    ("max_step", "corresponding_wall_time", "summary_metadata"),
+)
+class ScalarMetadata(_ScalarMetadata):
+  """Metadata about a scalar time series for a particular run and tag.
+
+  Attributes:
+    max_step: The largest step for this scalar time series; an integer.
+    corresponding_wall_time: The wall time of the datum with the largest
+      step in this time series, as `float` seconds since epoch.
+    summary_metadata: A `summary_pb2.SummaryMetadata` value.
+  """
+  pass
+
+
+_ScalarDatum = collections.namedtuple(
+    "ScalarDatum", ("step", "wall_time", "value")
+)
+class ScalarDatum(_ScalarDatum):
+  """A single datum in a scalar time series for a run and tag.
+
+  Attributes:
+    step: The global step at which this datum occurred; an integer. This
+      is a unique key among data of this time series.
+    wall_time: The real-world time at which this datum occurred, as
+      `float` seconds since epoch.
+    value: The scalar value for this datum; a `float`.
+  """
+  pass
+
+
+class RunTagFilter(object):
+  """Filters data by run and tag names."""
+
+  def __init__(self, runs=None, tags=None):
+    """Construct a `RunTagFilter`.
+
+    A time series passes this filter if both its run *and* its tag are
+    included in the corresponding whitelists.
+
+    Order and multiplicity are ignored; `runs` and `tags` are treated as
+    sets.
+
+    Args:
+      runs: Collection of run names, as strings, or `None` to admit all
+        runs.
+      tags: Collection of tag names, as strings, or `None` to admit all
+        tags.
+    """
+    self._runs = None if runs is None else frozenset(runs)
+    self._tags = None if tags is None else frozenset(tags)
+
+  @property
+  def runs(self):
+    return self._runs
+
+  @property
+  def tags(self):
+    return self._tags
+
+  def test(self, run, tag):
```

Comment:
Optional, but you might consider making these separate utilities for implementations of the provider API (and perhaps initially private) rather than methods on the classes.  They don't actually use any private fields, and that keeps the API surface more minimal.

Reply:
That’s probably a good idea. Will do.